src/Tools.js

```diff
@@ -60,5 +60,33 @@ class Tools {
 		}
 		return $result;
 	}
+
+	/**
+	 * Use jQuery objects (or strings) as i18n message parameters,
+	 * without having to parse translator-supplied strings as HTML.
+	 *
+	 * @param {string} msg The i18n message name.
+	 * @param {jQuery|string} params jQuery objects or strings to insert into the message.
+	 * @return {string}
+	 */
+	static i18nHtml( msg, ...params ) {
+		const
+			// Temporary marker string that will be replaced (with an integer appended).
+			// Must not appear in messages.
+			marker = '@@I18N',
+			// Create a list of parameter markers.
+			replacements = Object.keys( params ).map( function ( k ) { return marker + k; } );
+		// Insert the markers into the message, in place of $1, $2, etc.
+		let outputMsg = mw.msg.apply( null, [ msg ].concat( replacements ) );
+		// Replace each marker in the message with the actual HTML.
+		replacements.forEach( function ( rep, idx ) {
+			const param = params[ idx ],
+				// Parameters can be string or jQuery.
+				val = typeof param === 'string' ? param : param[ 0 ].outerHTML;
```

Comment:
Are we not risking the same problem we had before, where raw `<script>...</script>` in translated messages are actually inserted as HTML even if they're strings? They only difference between those two is that one would insert the raw string (that could be html) and one would insert the string representing outer html; both would then be parsed as HTML when injected into the output.

Am I missing something?

Reply:
I think the way to "fix" this is to do the same trick we do in `bidiIsolate` -- instead of outputting a string in this method that is then `.append( ... )` into jquery (which could mean we're injecting HTML we aren't controlling) -- we have this method always output either a jQuery object (parsed) or a jQuery.text() (which is always safe)

That would mean that whenever we do `.append( Tools.i18nHtml(...) )` we can always trust that we're appending safe html.

Reply:
You're quite right. I've replaced the string that's injected with `$( '<div>' ).text( param ).html()`, which escapes it. We could of course use `mw.html.escape()` but that's not available in tests; do you think it's worth mocking it? It seems easier to use jquery but I don't really mind.

The thing about string parameters here is that we don't ever actually use them at the moment. I only added support for them because it seems like it'd be nice to introduce this sort of thing into jquery.i18n (oh yes, that's the other reason to avoid mw.html).